game/src/core/hud/Inventory/InventorySlotSource.java

```diff
@@ -0,0 +1,88 @@
+package core.hud.Inventory;
+
+import com.badlogic.gdx.scenes.scene2d.Actor;
+import com.badlogic.gdx.scenes.scene2d.InputEvent;
+import com.badlogic.gdx.scenes.scene2d.ui.Window;
+import com.badlogic.gdx.scenes.scene2d.utils.DragAndDrop;
+
+import core.Game;
+import core.components.PositionComponent;
+
+public class InventorySlotSource extends DragAndDrop.Source {
+    private final InventorySlot inventorySlot;
+    private final DragAndDrop dragAndDrop;
+
+    /**
+     * Creates a new inventory slot source with the given slot. Source gets removed from the slot
+     * when dropped.
+     *
+     * @param slot The slot to create the source for.
+     * @param dragAndDrop The drag and drop object to use.
+     */
+    public InventorySlotSource(InventorySlot slot, DragAndDrop dragAndDrop) {
+        super(slot.getInventoryItem());
+        this.inventorySlot = slot;
+        this.dragAndDrop = dragAndDrop;
+    }
+
+    @Override
+    public DragAndDrop.Payload dragStart(InputEvent event, float x, float y, int pointer) {
+        DragAndDrop.Payload payload = new DragAndDrop.Payload();
+
+        Actor actor = getActor();
+        if (actor == null) {
+            return null;
+        }
+        // Removes description from slot
+        // this is used because dragStart somehow gets called before touchDragged
+        ((InventoryDescriptionListener) inventorySlot.getListeners().get(1))
+                .touchDragged(new InputEvent(), 0, 0, 0);
+
+        payload.setDragActor(actor);
+        // add the drag actor to the stage
+        InventoryGUI.getInstance().getInventoryWindow().getParent().addActor(actor);
+        dragAndDrop.setDragActorPosition(actor.getWidth() / 2, -actor.getHeight() / 2);
+        return payload;
+    }
+
+    @Override
+    public void dragStop(
+            InputEvent event,
+            float x,
+            float y,
+            int pointer,
+            DragAndDrop.Payload payload,
+            DragAndDrop.Target target) {
+        Window inv = InventoryGUI.getInstance().getInventoryWindow();
+        InventoryItem itemActor = (InventoryItem) payload.getDragActor();
+
+        if (target == null) {
+            // if the item is dropped outside the inventory, drop it on the ground
+            if (itemActor.getX() < inv.getX() - 40
+                    || itemActor.getX() > inv.getX() + inv.getWidth() - 20
+                    || itemActor.getY() < inv.getY() - 40
+                    || itemActor.getY() > inv.getY() + inv.getHeight() - 20) {
+                PositionComponent positionComponent =
+                        (PositionComponent)
+                                Game.getHero()
```

Comment:
@AMatutat noch ein punkt für die game-api: ich weiss, ich habe das optional angeschleppt. aber optional macht in java nur so richtig sinn, wenn in streams eingesetzt. bei `Game.hero()` sehe ich den stream-einsatz nicht wirklich. sollten wir nochmal überlegen, hier den hero direkt rauszugeben und ggf. das `orElseThrow` in der Methode selbst zu erledigen, d.h. dabei die exception ggf. zu werfen? allerdings sollte sich ein spiel auch ohne helden starten lassen (okay, da passiert ausser roamenden mobs nicht viel) und wir sollten auch den use-case "multiplayer" im kopf behalten.

Reply:
@cagix Ggf. auch beide "Flavour" /s parallel anbieten, Game.(get)Hero() einmal als Stream/Optional und einmal direkt/konkret (unwrapped). Damit es einheitlich wäre, aber bei allen infrage kommenden Attributen so handhaben.

Reply:
@tgrothe ich glaube, das sollte analog zu den entitäten gemacht werden.

Reply:
@cagix Aber, wenn man erst `Optional.ofNullable(hero)` aufrufen würde und dann `orElseThrow`, hätte man nichts gewonnen, da der Aufrufer dann immer noch auf eine Exception testen müsste ... deshalb stattdessen tatsächlich nur `return hero` und möglicherweise ein `null`-Check auf der anderen Seite. ... Und unabhängig davon, wie sich nun entschieden werden wird, sollte alles analog zu den Entities geschehen, ja.

Reply:
@tgrothe Die Diskussion ist hier an der Stelle ungeeignet. Eigentlich war das nur mein Hinweis an @AMatutat, das mit in das richtige Ticket zu nehmen (mein Luxus als PL). 

Letztlich sollte man eine Liste/Menge oder einen Stream rausgeben - dann kann der Aufrufe damit anfangen, was er möchte. Nur ein einzelnes Optional ist halt in diesem Fall nicht so wahnsinnig hilfreich. Und ich glaube, der Aufrufer muss/möchte gar nichts "fangen" in diesem speziellen Fall: Wenn es keinen Hero gibt, geht es nicht weiter, da kann ich als Aufrufer nichts beheben bzw. an der Situation ändern.

Reply:
@cagix ja, verstanden. ggf. an anderer stelle weiterdiskutieren.

Reply:
> Eigentlich war das nur mein Hinweis an @AMatutat, das mit in das richtige Ticket zu nehmen (mein Luxus als PL).

Öyy....

Reply:
Ist übertragen Herr PL.
Noch ein Kaffee? :P

Reply:
Hihi, gern! Bitte "Sträters ziemlich guten Kaffee", ohne Milch, ohne Zucker.

Reply:
![Alt Text](https://tenor.com/bqjTl.gif)

eigentlich fehlt jetzt noch ein Einfüllgeräusch, dann wäre es ein perfekter virtueller Kaffee ;)